ENH Adds TargetEncoder

thomasjpfan opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Closes https://github.com/scikit-learn/scikit-learn/pull/5853
Closes https://github.com/scikit-learn/scikit-learn/pull/9614
Supersedes https://github.com/scikit-learn/scikit-learn/pull/17323
Fixes or at least related to https://github.com/scikit-learn/scikit-learn/issues/24967


#### What does this implement/fix? Explain your changes.
This PR implements a target encoder which uses CV during `fit_transform` to prevent the target from leaking. `transform` uses the the target encoding from all the training data. This means that `fit_transform()` != `fit().transform()`. 

The implementation uses Cython to learn the encoding which provides a 10x speed up compared to using a pure Python+NumPy approach. Cython is required because many encodings are learn during cross validation in `fit_transform`.

#### Any other comments?
The implementation uses the same scheme as cuML's [TargetEncoder](https://docs.rapids.ai/api/cuml/stable/api.html#cuml.preprocessing.TargetEncoder.TargetEncoder), which they used to win [Recsys2020](https://medium.com/rapids-ai/winning-solution-of-recsys2020-challenge-gpu-accelerated-feature-engineering-and-training-for-cd67c5a87b1f).

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

thomasjpfan committed:
ENH Adds Target Regression Encoder

thomasjpfan committed:
DOC Adds pr number

thomasjpfan committed:
DOC Adds pr number

thomasjpfan committed:
DOC Adds example

thomasjpfan committed:
DOC Fixes example link

thomasjpfan committed:
FIX Use fancy indexing

thomasjpfan committed:
FIX Fix issue with 32bit

glemaitre commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -0,0 +1,214 @@
+
+        The used categories can be found in the ``categories_`` attribute.
+
+    smooth : float, default=30.0
```
From the dirty cat implementation and the paper, it seems that you can compute an empirical value (lambda_n / Eq. 8). It would be better to provide an "auto" version if it works.

I am wondering if we should not name this `shrinkage`. I am even wondering if really want to expose the parameter?

thomasjpfan replied:
I kind of like the semantics of `smooth`. For example, lets say there are 40 samples in training and `smooth=10.0`, then we take 20% of the target mean and 80% of the global mean. (This is the same semantics as cuml's TargetEncoder)

In any case, I can get behind an auto option using an empirical value or not exposing it at all.

thomasjpfan replied:
I updated this PR with a `smooth="auto"` that does exactly what the Micci-Barreca paper does (Equation 6). I kept the name "smooth", because it is controlling the `m` parameter from Equation 8 in the paper.

betatim commented on examples/preprocessing/plot_target_regressor_encoder.py:
```diff
@@ -0,0 +1,121 @@
+
+.. currentmodule:: sklearn.preprocessing
+
+The :class:`TargetRegressorEncoder` uses target statistics conditioned on
+the categorical features for encoding. In this example, we will compare
```
```suggestion
The :class:`TargetRegressorEncoder` uses the value of the target to encode each
categorical feature. In this example, we will compare
```

This is a bit more generic than the current sentence, but I think if you don't already know how the target encoder works, you won't learn it from the current sentence. "target statistics conditioned on
the categorical features for encoding" is pretty dense, so I think a simpler sentence that gives the reader a vague idea of what is happening is nicer.

betatim commented on examples/preprocessing/plot_target_regressor_encoder.py:
```diff
@@ -0,0 +1,121 @@
+
+The :class:`TargetRegressorEncoder` uses target statistics conditioned on
+the categorical features for encoding. In this example, we will compare
+:class:`TargetRegressorEncoder`, :class:`OrdinalEncoder`, and dropping the
```
```suggestion
three different approaches for handling categorical features: :class:`TargetRegressorEncoder`, :class:`OrdinalEncoder`, and dropping the
```

betatim commented on examples/preprocessing/plot_target_regressor_encoder.py:
```diff
@@ -0,0 +1,121 @@
+The :class:`TargetRegressorEncoder` uses target statistics conditioned on
+the categorical features for encoding. In this example, we will compare
+:class:`TargetRegressorEncoder`, :class:`OrdinalEncoder`, and dropping the
+category on a wine review dataset.
```
```suggestion
category.
```
Is it important that we use a wine reviews dataset? To me it seems like not really, and it is the next thing the reader sees.

betatim commented on examples/preprocessing/plot_target_regressor_encoder.py:
```diff
@@ -0,0 +1,121 @@
+
+X = df[numerical_features + categorical_features]
+y = df["points"]
+X.nunique().sort_values(ascending=False)
```
What does this do/why do I need to do this? Compared to the other lines of code in this cell which are fairly self-explanatory, I have no idea what this line does and why it is here. If subsetting the features is worth explaining, this is for sure worth explaining :D

betatim commented on examples/preprocessing/plot_target_regressor_encoder.py:
```diff
@@ -0,0 +1,121 @@
+# =======================================================
+# Dropping the categorical features
+# ---------------------------------
+# As a basline, we construct a pipeline where the categorical features are
```
```suggestion
# As a baseline, we construct a pipeline where the categorical features are
```

betatim commented on examples/preprocessing/plot_target_regressor_encoder.py:
```diff
@@ -0,0 +1,121 @@
+    ]
+)
+
+reg_drop_cats = Pipeline(
```
```suggestion
without_categories = Pipeline(
```
Started thinking about the variable name because I tried to work out what the `reg_` meant. Couldn't work it out and then ended up with this suggestion. WDYT? (Shortened things like `cats`, `cnt`, etc are a pet peeve of mine. I prefer the spelt out version as it has nearly no cost but lots of readability improvement.)

thomasjpfan replied:
I'm okay with having better names.

betatim replied:
Do you like the ones I suggested? Otherwise I will think up some more options that we can choose from/inspire a good name.

betatim commented on examples/preprocessing/plot_target_regressor_encoder.py:
```diff
@@ -0,0 +1,121 @@
+
+# %%
+# Here we train and use the root mean squared error to evalute the baseline
+# model:
```
Is there a short sentence for why RMSE is a good/sensible choice? I've not looked at the data but I could imagine that the score/points/rating the reviewers assign are like `1`, `2`, `3`, etc, not truly continuous values. As a reader I'm now wondering why it is called out that the obvious(?) choice of RMSE is used for evaluating the performance of the pipeline, but we didn't talk about the default parameters of the gradient boosting model.

So if there is a nice short sentence to explain it, that would be great. Otherwise, I understand we need some kind of sentence for this cell :-/ but it would be nice to use it to add to what the code already says (ideally explaining why the code is what it is).

thomasjpfan replied:
The unique points are:

```python
array([ 96.,  95.,  94.,  90.,  91.,  86.,  89.,  88.,  87.,  93.,  92.,
        85.,  84.,  83.,  82.,  81., 100.,  99.,  98.,  97.,  80.])
```

I wanted to model this as regression, because there is an intrinsic order to the points. The other way to keep the order semantics would be Ordinal Regression, but I do not think anyone has look into it for target encoders.

I agree, this section can be expanded a little more.

**Aside**: [Here is a fun library](https://raschka-research-group.github.io/coral-pytorch/) that does Ordinal Regression.

thomasjpfan replied:
I removed the mention of RMSE. My aim is to focus the narrative on encoding and less on the choice of metric. Choosing a metric for evaluation is a much bigger topic and I do not want to squeeze it into one sentence.

betatim replied:
Sounds good.

betatim commented on examples/preprocessing/plot_target_regressor_encoder.py:
```diff
@@ -0,0 +1,121 @@
+# %%
+# We modify the original pipeline to use the ordinal categorical preprocessing:
+reg_ordinal = reg_drop_cats.set_params(prep__cat=cat_prep)
+reg_ordinal
```
```suggestion
ordinal_encoded
```
trying to follow the pattern of the above variable name suggestion.Only applies if we change the `reg_drop_cats`.

betatim commented on examples/preprocessing/plot_target_regressor_encoder.py:
```diff
@@ -0,0 +1,121 @@
+from sklearn.preprocessing import TargetRegressorEncoder
+
+reg_target = reg_ordinal.set_params(prep__cat=TargetRegressorEncoder())
+reg_target
```
```suggestion
target_encoded
```
(as above re: name suggestions)

betatim commented on examples/preprocessing/plot_target_regressor_encoder.py:
```diff
@@ -0,0 +1,121 @@
+# The :class:`TargetRegressorEncoder` further improves the RMSE:
+reg_target.fit(X_train, y_train)
+reg_target_rmse = mean_squared_error(y_test, reg_target.predict(X_test), squared=False)
+print(f"RMSE with target encoding: {reg_target_rmse:.4}")
```
Can we have some kind of wrap up/conclusion? Something like "As expected the performance improves when using the additional categorical features, compared to using only the price of the wine to predict its rating. The `TargetRegressorEncoder` further improves the performance because it does not assign an order to the categorical features, which in reality do not have an order."

I don't know if this is the reason why the performance improves, so just a guess. Does anyone know "the" reason?

betatim commented:
> Do we want to have `Regressor` in the naming since we could make the encoder work on both regression and classification problem?

To have a encoder that works for both classification and regression I think we'd have to detect the type of problem based on the vaules of `y`. How good is the automatic detection of "regression vs classification"? In addition it feels like the code for classification, in particular multi-class classification, would be more complicated than for the regression case (you can't use `np.mean(y)`).

---

I think the way this works is that when performing the change from "categories" to "encoded categories" `_BaseEncoder._transform` returns a version of `X` where a column representing a categorical feature is replaced by a column of integers. This is then translated to the actual encoded value in `_transform_X_int`. On top we have the complexity of `fit().transform()` and `fit_transform()` not doing exactly the same thing. It took a lot of going back and forth and working out what `_BaseEncoder` does to get to that understanding. 

After reading the code and example once my big picture thoughts are:
* nice!
* this looks like it should work,
* can we have more doc strings and comments that explain why something is the way it is
* both doc string and naming stuff should be done after code structure, etc
* right now sample weights aren't supported, would it be easy to add that?

What is your plan? Is it ready to go or does it need more tweaking?

glemaitre commented:
> How good is the automatic detection of "regression vs classification"?

`type_of_target` is quite good and would do the job. 

> In addition it feels like the code for classification, in particular multi-class classification, would be more complicated than for the regression case (you can't use np.mean(y)).

`dirty_cat` implements both but this is not the same branch of code: https://github.com/dirty-cat/dirty_cat/blob/main/dirty_cat/_target_encoder.py

betatim commented:
What can the `TargetEncoder` from dirty cat not do/what is the advantage of making yet another implementation over taking it verbatim? For example from the constructor args of it, it seems it supports regression, binary and multi-class classification. It seems like an excellent starting point

glemaitre commented:
From the dev meeting, we thought that could put aside the multiclass on the side for the moment and support binary classification and regression. We need to make sure that the name of the encoder reflects that but we don't have to support all possible classification and regression problems at first.

I will make a review having those points in mind.

glemaitre commented:
> What can the TargetEncoder from dirty cat not do/what is the advantage of making yet another implementation over taking it verbatim?

Since `OneHotEncoder` and `OrdinalEncoder` are working for all problems, it could come as a surprise to our user to have to select the right encoder type. There is also the precedent of `categorical_encoder`, `cuml`, `dirty_cat` (for the one that I am aware of) that are exposing a single encoder for all usage.

betatim commented:
> Since `OneHotEncoder` and `OrdinalEncoder` are working for all problems, it could come as a surprise to our user to have to select the right encoder type.

I agree. Having one encoder for all types of problems is nicer than having to choose.

My question was "Why not take the dirty_cat implementation of `TargetEncoder` (as a starting point) for the implementation in scikit-learn?"

thomasjpfan commented:
Here are the key differences between this PR and [dirty_cat's version](https://github.com/dirty-cat/dirty_cat/blob/main/dirty_cat/_target_encoder.py):

1. This PR takes advantage of the existing code in `_BaseEncoder`, which simplifies the implementation, while dirty_cat implements integer encoding from scratch.
2. This PR does cross validation in `fit_transform` while dirty cat does not. The cross validation is required to prevent the target from leaking during training.
3. Lastly this PR's implementation is much faster because of the Cython:

```python
from sklearn.preprocessing import OrdinalEncoder
from sklearn.preprocessing import TargetRegressorEncoder
from dirty_cat import TargetEncoder as DirtyCatTargetEncoder
import numpy as np

rng = np.random.default_rng()
n_samples, n_features = 500_000, 20
X = rng.integers(0, high=30, size=(n_samples, n_features))
y = rng.standard_normal(size=n_samples)
```

```python
%%timeit
_ = TargetRegressorEncoder().fit_transform(X, y)
# 3.37 s ± 132 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%%timeit
_ = DirtyCatTargetEncoder().fit_transform(X, y)
# 9.9 s ± 220 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

This PR is faster even when it's `fit_transform` learns 6 encodings: five for `cv=5` and one more encoding over the whole dataset. `dirty_cat` only learns one encoding for the whole dataset because it does not do cross validation.

thomasjpfan commented:
> What is your plan? Is it ready to go or does it need more tweaking?

It's mostly to decide on how we want to extend the API for classification targets. Currently, this PR is the minimum requirement for regression targets. The core computation in this PR can be extended to classification without too much trouble.

> type_of_target is quite good and would do the job.

I do not like how `type_of_target` will treats floats as classification targets. For example:

```python
import numpy as np
from sklearn.utils.multiclass import type_of_target

type_of_target(np.asarray([1.0] * 10 + [2.0] * 30 + [4.0] * 10 + [5.0]))
# 'multiclass'
```

I prefer two more explicit options:

1. `TargetRegressorEncoder` and  `TargetClassificationEncoder`
2. `TargetEncoder` with a `target_type` parameter to switch between regression and classification.

I went with option 1 in this PR, but I am okay with either option.

thomasjpfan committed:
ENH Adds support for binary classification

thomasjpfan committed:
TST Adds test to check target_type_

thomasjpfan commented:
After thinking about it a little more, I am okay with just inferring the target type with `type_of_target`. I updated this PR to support both binary classification and regression.

thomasjpfan committed:
DOC Update whats new with better wording

thomasjpfan committed:
DOC Update docs about target encoder itself

thomasjpfan committed:
TST Update names of tests

thomasjpfan committed:
DOC better names for variables

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into target_regression_encoder_v2_pr

thomasjpfan committed:
ENH Adds target_type parameter

thomasjpfan commented:
From the example, the `y.unique()` was:

```python
array([ 96.,  95.,  94.,  90.,  91.,  86.,  89.,  88.,  87.,  93.,  92.,
        85.,  84.,  83.,  82.,  81., 100.,  99.,  98.,  97.,  80.])
```

`type_of_target` inferred this as multi-class, but I wanted to model it as regression. I think this use case requires a `target_type` parameter to be added to `TargetEncoder`.

thomasjpfan committed:
TST Fixes test failures

thomasjpfan committed:
DOC Remove mention on RMSE

thomasjpfan committed:
DOC Adds concluding statement in example

thomasjpfan committed:
CLN Refactor names

betatim commented:
A thought on naming: `TargetEncoder` names it after the algorithm it uses (lots of precedent for this in the library), `NominalEncoder` names it after when to use it, similar to `OrdinalEncoder`, which is nice because the naming reflects that they form a "pair".

I think I prefer `TargetEncoder` because "nominal categories" strikes me as a bit of obscure knowledge. At least I didn't know what it was until educating myself. My first association for "nominal" is "everything is nominal", as in "where it should be".

betatim commented on doc/modules/preprocessing.rst:
```diff
@@ -830,6 +830,54 @@ lexicon order.
 
+.. _target_encoder:
+
+Target Regressor Encoder
+------------------------
```
```suggestion
Target Encoder
--------------
```

betatim commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -0,0 +1,249 @@
+    See Also
+    --------
+    OrdinalEncoder : Performs an ordinal (integer) encoding of the categorical features.
+    OneHotEncoder : Performs a one-hot encoding of categorical features.
```
We should cross-reference `TargetEncoder` from onehot's and ordinal's docstring as well

thomasjpfan committed:
DOC Improves user guide

thomasjpfan committed:
CLN Cleaner implementation

thomasjpfan committed:
ENH Adds auto for smoothing

thomasjpfan committed:
CLN Address comments

amueller commented on doc/modules/preprocessing.rst:
```diff
@@ -830,6 +830,72 @@ lexicon order.
+
+.. currentmodule:: sklearn.preprocessing
+
+The :class:`TargetEncoder` uses target mean groupby the
```
I feel saying "per value of the category feature" is going to be more clear? Or "grouped by"? `groupby` is not a common usage within ml, right?

amueller commented on doc/modules/preprocessing.rst:
```diff
@@ -830,6 +830,72 @@ lexicon order.
+categorical feature for encoding the categories [PAR]_ [MIC]_. This encoding
+scheme is useful with categorical features with high cardinality, where one hot
+encoding would inflate the feature space making it more expensive for a
+downstream model to process. A classical example of high cardinality categories
```
and harder to generalize.

thomasjpfan committed:
DOC Update example title

amueller commented on doc/modules/preprocessing.rst:
```diff
@@ -830,6 +830,72 @@ lexicon order.
+scheme is useful with categorical features with high cardinality, where one hot
+encoding would inflate the feature space making it more expensive for a
+downstream model to process. A classical example of high cardinality categories
+are location based such as zip code or region. For the binary classification
```
```suggestion
are location based such as zip code or region. For a binary classification
```
so is this doing just regression or regression and binary classification for now?

thomasjpfan committed:
DOC Adds comment to point to equation

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into target_regression_encoder_v2_pr

thomasjpfan commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -0,0 +1,263 @@
+from ..utils._param_validation import Interval, StrOptions
+
+
+class TargetEncoder(OneToOneFeatureMixin, _BaseEncoder):
```
I'm planning to have the same class do multi-class.

thomasjpfan committed:
Apply suggestions from code review

Co-authored-by: Andreas Mueller <t3kcit@gmail.com>

thomasjpfan committed:
DOC Only say classification

thomasjpfan committed:
DOC Update examples and docstring

amueller commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -0,0 +1,304 @@
+)
+@pytest.mark.parametrize("smooth", [4.0, "auto"])
+def test_custom_categories(X, categories, smooth):
+    """custom categoires with unknown categories that are not in training data."""
```
```suggestion
    """custom categories with unknown categories that are not in training data."""
```

amueller commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -0,0 +1,304 @@
+
+    if target_type == "binary":
+        y_int = rng.randint(low=0, high=2, size=n_samples)
+        target_names = np.asarray(["cat", "dog"], dtype=object)
```
``np.asarray`` is equivalent to ``np.array`` here right?

amueller commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -0,0 +1,304 @@
+        y_int = rng.uniform(low=-10, high=20, size=n_samples)
+        y_input = y_int
+
+    # compute encodings for all data to validate `transform`
```
I guess I would do that further down close to where we actually validate transform, but just a nit...

amueller commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -0,0 +1,304 @@
+def test_use_regression_target():
+    """Custom target_type to avoid inferring the target type."""
+    X = np.asarray([[0, 1, 0, 1, 0, 1]]).T
+    # type_of_target would be 'multiclass'
```
that seems bad? at least once we support multiclass this would be a silent unexpected behavior, right?

amueller replied:
I think an assumption of type_of_target is that any model only supports classification or regression, but here we support both?

amueller commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -0,0 +1,304 @@
+    X_pandas = enc_pandas.fit_transform(X_df, y)
+
+    assert_allclose(X_pandas.to_numpy(), X_default)
+    assert_array_equal(enc_pandas.get_feature_names_out(), X_pandas.columns)
```
can we also check the actual feature names / column names?

amueller commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -0,0 +1,304 @@
+
+    if to_pandas:
+        pd = pytest.importorskip("pandas")
+        # convert second feature to a object
```
```suggestion
        # convert second feature to an object
```

amueller commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -0,0 +1,304 @@
+)
+@pytest.mark.parametrize("smooth", ["auto", 4])
+def test_constant_target_and_feature(y, smooth):
+    """Check edge case where feature and target is constant."""
```
what happens in the string case for classification? the single class is the 0th class I assume? is it worth testing?

thomasjpfan replied:
Yea, it will become 0. I think it's worth checking.

amueller commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -0,0 +1,263 @@
+
+        - `"auto"` : Type of target is inferred with
+          :func:`~sklearn.utils.multiclass.type_of_target`
+        - `"continuous"` : Continious target
```
```suggestion
        - `"continuous"` : Continuous target
```

amueller commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -0,0 +1,263 @@
+        - `"binary"` : Binary target
+
+    smooth : "auto" or float, default="auto"
+        The amount of mixing the categorical encoding with the global target mean. A
```
```suggestion
        The amount of mixing of the categorical encoding with the global target mean. A
```

thomasjpfan committed:
DOC Imrove example

thomasjpfan committed:
DOC Clarify fit_transform vs fit.transform

thomasjpfan committed:
CLN Address comments

thomasjpfan committed:
DOC Adds more code comments

thomasjpfan committed:
CLN Code formatting

thomasjpfan committed:
CLN Move test closer to comment

thomasjpfan committed:
DOC Adds more details aboue equations

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into target_regression_encoder_v2_pr

glemaitre reviewed (commented):
This is a partial review.

thomasjpfan commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -0,0 +1,263 @@
+    target_type_ : str
+        Type of target.
+
+    encoding_mean_ : float
```
I do not think there is a leak in `fit_transform`, because `encoding_mean_` is not used to encode anything in `fit_transform`. `fit_transform` uses a CV to computes a mean for every "training split" to encode the "testing split". `encoding_mean_` is only used to encode data in `transform`.

thomasjpfan commented on examples/preprocessing/plot_target_encoder.py:
```diff
@@ -0,0 +1,105 @@
+# be a reviewer:
+from sklearn.datasets import fetch_openml
+
+wine_reviews = fetch_openml(data_id=42074, as_frame=True, parser="liac-arff")
```
When running:

```python
from sklearn.datasets import fetch_openml

wine_reviews = fetch_openml(data_id=42074, as_frame=True, parser="pandas")

# pandas.errors.ParserError: Error tokenizing data. C error: Expected 10 fields in line 2, saw 15
```

I have not looked too deep into what is causing the issue.

glemaitre replied:
I have to look at this one. I will raise an issue to track it.

glemaitre replied:
Some kind of fixes here: https://github.com/scikit-learn/scikit-learn/pull/25511

thomasjpfan commented on examples/preprocessing/plot_target_encoder.py:
```diff
@@ -0,0 +1,105 @@
+        "one_hot_encoder",
+        OneHotEncoder(handle_unknown="ignore", max_categories=50, sparse_output=False),
+    ),
+    ("target", TargetEncoder(target_type="continuous")),
```
Since the target is discrete and integers:

```python
y.unique()
# array([ 96,  95,  94,  90,  91,  86,  89,  88,  87,  93,  92,  85,  84,
#        83,  82,  81, 100,  99,  98,  97,  80])
```

`type_of_target` is treating this as a multiclass problem. For this example, I wanted to model this as a regression.

thomasjpfan commented on sklearn/preprocessing/_target_encoder_fast.pyx:
```diff
@@ -0,0 +1,130 @@
+        double[::1] sums = np.empty(max_cats, dtype=np.float64)
+        double[::1] counts = np.empty(max_cats, dtype=np.float64)
+
+    for j in range(n_features):
+        n_cats = n_categories[j]
```
I think it is possible. I would need to make tweak the loop such that it does not require the gil and I likely need `X_int` to be F-ordered for data locality.

I think it's something to experiment with as a follow up.

thomasjpfan commented on sklearn/preprocessing/_target_encoder_fast.pyx:
```diff
@@ -0,0 +1,130 @@
+    for j in range(n_features):
+        n_cats = n_categories[j]
+
+        for i in range(n_cats):
+            means[i] = 0.0
+            counts[i] = 0.0
+            sum_of_squared_diffs[i] = 0.0
```
I had a version where I created different sized arrays in the loop and it was much slower compared to reusing an array with `max_n_cats`. From memory, I think it was about ~40% slower.

thomasjpfan committed:
CLN Address comments

thomasjpfan commented on sklearn/preprocessing/_target_encoder_fast.pyx:
```diff
@@ -0,0 +1,130 @@
+            if X_int_tmp == -1:
+                continue
+            counts[X_int_tmp] += 1.0
+            means[X_int_tmp] += (y[i] - means[X_int_tmp]) / counts[X_int_tmp]
```
Yea, I was thinking about overflow. I'll switch to the sum. Overflowing a float64 from the target feels like an edge case.

glemaitre commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -0,0 +1,270 @@
+        """Fit a target encoding with all the data."""
+        from ..preprocessing import LabelEncoder  # avoid circular import
+
+        self._validate_params()
```
It would be better to put it in the public function instead.

glemaitre commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -0,0 +1,270 @@
+
+        return X_int, X_mask, y, n_categories
+
+    def _transform_X_int(self, X_out, X_int, X_invalid, indices, encodings, y_mean):
```
```suggestion
    @staticmethod
    def _transform_X_int(X_out, X_int, X_invalid, indices, encodings, y_mean):
```

glemaitre commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -0,0 +1,270 @@
+        (in order of the features in `X` and corresponding with the output
+        of :meth:`transform`).
+
+    target_type_ : str
```
Somehow I think that we should add some more documentation highlighting that target type inference can fail.

thomasjpfan replied:
I added a note to the `target_type` parameter about how the automatic target type inference may not be desired.

glemaitre commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -0,0 +1,308 @@
+    assert_array_equal(enc_pandas.get_feature_names_out(), X_pandas.columns)
+
+
+@pytest.mark.parametrize("to_pandas", [True, False])
```
I am certainly biased but I liked to use `_convert_container` from the `testing` utils.

thomasjpfan replied:
For this specific test, if `to_pandas=True`, then the second feature is converted into a strings improve test coverage. I do not see a way to add `_convert_container` without complicating the logic in:

https://github.com/thomasjpfan/scikit-learn/blob/acb4b2eb4fea92ef0ee6507eb3c5de713a9bafd0/sklearn/preprocessing/tests/test_target_encoder.py#L235-L244

I guess it can become the following:

```python
    if constructor_name == "dataframe":
        # convert second feature to an object
        X_input = _convert_container(
            {
                "feat0": X_int[:, 0],
                "feat1": np.array(["cat", "dog"], dtype=object)[X_int[:, 1]],
            },
            constructor_name,
        )
        # "snake" is unknown
        X_test = _convert_container(
            {"feat0": X_test[:, 0], "feat1": ["dog", "cat", "snake"]}, constructor_name
        )
```

But I feel it adds a little bit too much indirection.

glemaitre replied:
OK. Fine then.

glemaitre reviewed (commented):
I will look at the test tomorrow.

thomasjpfan committed:
TST Simplify test logic

thomasjpfan committed:
CLN Address comments

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into target_regression_encoder_v2_pr

thomasjpfan committed:
DOC Adds note about type_of_target inference

thomasjpfan committed:
ENH Use integers for counts

thomasjpfan committed:
STY Formatting

thomasjpfan committed:
STY Fix numpydoc linting error

thomasjpfan committed:
DOC Adds example about low and high smoothing parameters

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into target_regression_encoder_v2_pr

ogrisel commented:
Any idea why the `doc-min-dependencies` fails with:

```
Sphinx Warnings in affected files

    sklearn/preprocessing/_target_encoder.py:docstring of sklearn.preprocessing._target_encoder.TargetEncoder:127: WARNING: Unknown target name: "enc_high_smooth.encoding_mean".
    sklearn/preprocessing/_target_encoder.py:docstring of sklearn.preprocessing._target_encoder.TargetEncoder:127: WARNING: Unknown target name: "enc_high_smooth.encodings".
    sklearn/preprocessing/_target_encoder.py:docstring of sklearn.preprocessing._target_encoder.TargetEncoder:135: WARNING: Unknown target name: "enc_no_smooth.encodings". 
```

while the `doc` builds is fine

ogrisel commented on examples/preprocessing/plot_target_encoder.py:
```diff
@@ -0,0 +1,120 @@
+# the significance of the improvement.
+import pandas as pd
+
+pd.DataFrame(results).sort_values("root mean squared error", ascending=False).set_index(
```
```suggestion
_ = pd.DataFrame(results).sort_values("root mean squared error", ascending=False).set_index(
```

ogrisel commented on examples/preprocessing/plot_target_encoder.py:
```diff
@@ -0,0 +1,120 @@
+
+pd.DataFrame(results).sort_values("root mean squared error", ascending=False).set_index(
+    "preprocessor"
+).plot(kind="barh")
```
This plot has cut labels in the HTML rendering of the figure. I suspect with need the constrained layout. I am not sure how to do that directly via the pandas API though.

ogrisel commented on examples/preprocessing/plot_target_encoder.py:
```diff
@@ -0,0 +1,120 @@
+    ("drop", "drop"),
+    ("ordinal", OrdinalEncoder(handle_unknown="use_encoded_value", unknown_value=-1)),
+    (
+        "one_hot_encoder",
```
For the sake of consistency with the other preprocessor names:

```suggestion
        "one_hot",
```

This might also fix the cropped label problem as a side effect (without having to use the constrained layout).

ogrisel commented on examples/preprocessing/plot_target_encoder.py:
```diff
@@ -0,0 +1,120 @@
+# more information about the target, which the gradient boosted model at the end
+# of the pipeline can take advantage of. In this example, we evaluate with a single
+# test split to reduce the runtime. We recommend using cross-validation to check
+# the significance of the improvement.
```
The code wouldn't be significantly more complex if we had called `cross_validate` on the full dataset instead of calling `fit` / `predict` / `mean_squared_error` manually on a single train-test split. Don't you think we should rather do the right thing directly here?

["Show, don't tell"](https://en.wikipedia.org/wiki/Show,_don%27t_tell)

thomasjpfan replied:
Yea, that makes sense. I updated the PR to use `cross_validate`, but with `cv=3` to have a manageable runtime.

ogrisel commented on examples/preprocessing/plot_target_encoder.py:
```diff
@@ -0,0 +1,120 @@
+
+pd.DataFrame(results).sort_values("root mean squared error", ascending=False).set_index(
+    "preprocessor"
+).plot(kind="barh")
```
Also I think we should add a new section at the end of the example to explain why we did not use the native categorical feature support of `HistGradientBoostingRegressor` with the ordinal encoding in this example. The reason being that some categorical features have more than 256 unique values which is the maximum number of categories supported by `HistGradientBoostingRegressor`  per categorical feature.

```python
>>> for feature_name in categorical_features:
...     print(f"{feature_name}: {df[feature_name].nunique()}")
... 
country: 48
province: 455
region_1: 1236
region_2: 18
variety: 632
winery: 14810
```

We could have split the categorical features in two groups (high and low cardinality) to use the target encoder on the high cardinality categorical features and the native categorical support on the low cardinality features but we leave that as an exercise to the reader (unless you think it would be interesting to expand this example to implement the mixed strategy).

thomasjpfan replied:
I think it's worth including the mixed strategy. I expanded the PR to include it.

ogrisel commented on examples/preprocessing/plot_target_encoder.py:
```diff
@@ -0,0 +1,120 @@
+# predictive performance on the test set, dropping the categories perform the
+# worst and the target encoder performs the best. The target encoding provides
+# more information about the target, which the gradient boosted model at the end
+# of the pipeline can take advantage of. In this example, we evaluate with a single
```
I think the main explanation is that:

- dropping the categorical features makes the pipeline less expressive and underfitting as a result;
- the one-hot encoding scheme makes the pipeline overfitting as the larger number of features explodes with rare category occurrences that are correlated with the target by chance (on the training set only);
- the ordinal encoding imposes an arbitrary order to the features which are then treated as numerical values by the `HistGradientBoostingRegressor` model. Since this model groups numerical features in 256 bins per feature, many unrelated categories can be grouped together and as a result overall pipeline can underfit.
- when using the target encoder, the same binning happens, but since the encoded values are statistically ordered by marginal association with the target variable, the binning use by the `HistGradientBoostingRegressor` model makes sense and leads to good results: the combination of smoothed target encoding and binning works as a good regularizing strategy against overfitting while not limiting the expressiveness of the pipeline too much.

I think we should extend the plot to also display the training RMSE to confirm the above analysis and make sure it is correct (biggest train / test gap for OHE, train and test much closer for the presumably underfitting models, with the train RMSE constraining the performance). Ideally this would all be with error bars with `cross_validate` together with `return_train_score=True`.

ogrisel reviewed (commented):
Here is a pass of review focused on the example (it's nice to see such great differences w.r.t. the existing strategies by the way):

thomasjpfan committed:
Apply suggestions from code review

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

thomasjpfan committed:
STY Black formating

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into target_regression_encoder_v2_pr

thomasjpfan committed:
DOC Adds mixed encoder for high and low cardinality

thomasjpfan committed:
DOC Address comments

ogrisel commented on examples/preprocessing/plot_target_encoder.py:
```diff
@@ -0,0 +1,224 @@
+# --------------------
+# In this section, we display the results for all the encoders. When evaluating the
+# predictive performance on the test set, dropping the categories perform the
+# worst and the target encoders performs the best.
```
There is a problem in the HTML rendering of the first bullet item. It probably requires a blank line:

```suggestion
# worst and the target encoders performs the best. This can be explained
# as follows:
#
```

ogrisel commented on examples/preprocessing/plot_target_encoder.py:
```diff
@@ -0,0 +1,224 @@
+# - Due to the high cardinality and to reduce the training time, the one-hot
+#   encoding scheme uses `max_categories=20` which prevents the features from
+#   expanding too much, which can result in underfitting.
+# - The one-hot encoding scheme makes the pipeline overfitting as the larger
+#   number of features explodes with rare category occurrences that are correlated
+#   with the target by chance (on the training set only);
```
```suggestion
# - If we had not set `max_categories=20`, the one-hot encoding scheme would have
#   likely made the pipeline overfitting as the number of features explodes with rare
#   category occurrences that are correlated with the target by chance (on the training
#   set only);
```

ogrisel commented on examples/preprocessing/plot_target_encoder.py:
```diff
@@ -0,0 +1,224 @@
+#   the binning use by the :class:`~sklearn.ensemble.HistGradientBoostingRegressor`
+#   makes sense and leads to good results: the combination of smoothed target
+#   encoding and binning works as a good regularizing strategy against overfitting while
+#   not limiting the expressiveness of the pipeline too much.
```
I would rather move the analysis below the plot, to avoid having to scroll through the code of the plot when reading the analysis.

ogrisel reviewed (commented):
The new version of the example is really interesting. Thanks for the extra work. Here is some feedback on that and then I will review the actual implementation.

thomasjpfan committed:
ENH Make sure the colors align between graphs

thomasjpfan committed:
Apply suggestions from code review

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

thomasjpfan committed:
CLN Reformat section headers and move sections around

thomasjpfan committed:
CLN Try to work around sphinx 4.0.1

betatim commented on doc/modules/preprocessing.rst:
```diff
@@ -830,6 +830,72 @@ lexicon order.
+  :class:`TargetEncoder` uses a cross validation scheme in
+  :meth:`~TargetEncoder.fit_transform` to prevent leaking the target during training.
+  In :meth:`~TargetEncoder.fit_transform`, the data is split according to the `cv`
+  parameter. Categorical encodings are learned from split and used to encode the other
```
```suggestion
  parameter. Categorical encodings are learned from one split and used to encode the other
```

thomasjpfan committed:
CLN Address comments

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into target_regression_encoder_v2_pr

thomasjpfan committed:
CLN Better variable names

amueller commented:
Not 100% sure about using the same class for classification and regression but it's probably fine.

ogrisel commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -0,0 +1,304 @@
+                )
+            self._transform_X_ordinal(
+                X_out, X_ordinal, X_unknown_mask, test_idx, encodings, y_mean
+            )
```
In case of overlapping `test_idx`, e.g. when using `ShuffleSplit` with a larger number of splits, the last iterations will overwrite the previously encoded values.

Is this a problem, should we try to detect this case and warn the user? Or just document in the `cv` parameter.

What about cases where not all samples appear at lest once as `test_idx`? E.g. `cv=ShuffleSplit(n_splits=1, test_size=0.5)`? Reading this code I would have thought that `X_out` would hold uninitialized values (from the `np.empty_like` allocation above). However it does not seem to be the case when I tried but I am not sure I understand why...

ogrisel replied:
Actually, it can show it is a problem when disabling smoothing:

```python
>>> from sklearn.model_selection import ShuffleSplit
>>> from sklearn.preprocessing import TargetEncoder
>>> import numpy as np
>>> X = np.random.randint(0, 3, 30).reshape(-1, 1)
>>> TargetEncoder(cv=ShuffleSplit(n_splits=1, test_size=0.5), smooth=0).fit_transform(X, X.ravel() + np.random.randn(X.shape[0]))
array([[ 0.00000000e+000],
       [ 9.88131292e-324],
       [ 4.22543968e-001],
       [ 9.88131292e-324],
       [ 0.00000000e+000],
       [ 4.22543968e-001],
       [ 9.88131292e-324],
       [ 4.94065646e-324],
       [ 4.94065646e-324],
       [ 2.26479250e+000],
       [ 2.26479250e+000],
       [ 4.22543968e-001],
       [ 0.00000000e+000],
       [ 4.22543968e-001],
       [ 4.22543968e-001],
       [-5.53085205e-002],
       [-5.53085205e-002],
       [ 2.26479250e+000],
       [ 4.94065646e-324],
       [-5.53085205e-002],
       [ 4.94065646e-324],
       [ 4.22543968e-001],
       [ 4.22543968e-001],
       [ 4.94065646e-324],
       [ 9.88131292e-324],
       [ 2.26479250e+000],
       [ 2.26479250e+000],
       [ 9.88131292e-324],
       [ 9.88131292e-324],
       [ 0.00000000e+000]])
```

ogrisel replied:
Those `e-324` values look fishy.

thomasjpfan replied:
In this case, I think it's best to only allow integers for the `cv` parameter for the initial version. As you said, there are quite a few ways to get the wrong result with a custom CV.

ogrisel replied:
Fair enough, an alternative would be to check for non-overlapping validation folds but this can be expensive. Let's start with integer specified k-fold and we will see later if we need to expand the allowed parametrization for the `cv` parameter or not. But I agree this is probably a YAGNI.

ogrisel reviewed (commented):
I need to finish reading the paper but here is already a small pass of feedback.

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into target_regression_encoder_v2_pr

thomasjpfan committed:
ENH Restrict cv to integers

thomasjpfan committed:
DOC Improve docstring

thomasjpfan committed:
CLN Better variable names in test

thomasjpfan committed:
Apply suggestions from code review

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

thomasjpfan committed:
DOC Simplify notes

thomasjpfan committed:
DOC Adds diagram about cross validation

thomasjpfan committed:
DOC Explain fit vs fit_transform

thomasjpfan commented on doc/modules/preprocessing.rst:
```diff
@@ -830,6 +830,72 @@ lexicon order.
+  parameter. Categorical encodings are learned from one split and used to encode the other
+  split. Afterwards, a final categorical encoding is learned from all the data, which
+  is then used to encode data during :meth:`~TargetEncoder.transform`. This means that
+  `fit(X, y).transform(X)` does not equal `fit_transform(X, y)`.
```
I expanded on how `fit_transform` and `fit` by itself works, along with a diagram to explain how the CV process works in `fit_transform`.

thomasjpfan committed:
DOC Clarify fit

ogrisel reviewed (commented):
Here is another pass of review. The main problem I found is a likely unwanted leak of y_train into the shrinkage value when y_train is statistically dependent on the ordering of the training samples.

I therefore think we should probably using a shuffling internal CV by default as explained in the comment:

ogrisel commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -0,0 +1,308 @@
+        self._validate_params()
+        X_ordinal, X_known_mask, y, n_categories = self._fit_encodings_all(X, y)
+
+        if self.target_type_ == "continuous":
```
```suggestion
        # The cv splitter is voluntarily restricted to *KFold to enforce non
        # overlapping validation folds, otherwise the fit_transform output will
        # not be well-specified.
        if self.target_type_ == "continuous":
```

ogrisel reviewed (commented):
Other comment: as of now, TargetEncoder has no explicit support for missing values and in particular encoded with `np.nan` in the original data are target encoded as any other categories.

In the future, one might want to have an extra constructor parameter to specify which category label should be treated as a missing value marker and offer the possibility to encode them explicitly as `np.nan` in the output so that downstream models with native missing value handling can be leveraged instead of marginal target encoding of those values.

I think the current behavior is a reasonable default though, and maybe `np.nan` encoding for downstream models with native missing value support will be practically useless.

Still, I think the user guide and/or the docstring of this estimator should be explicit about the fact that `TargetEncoder` in this current form has no explicit missing value handling in general and `np.nan` or `None` values in the original input will be treated as another category label and target encoded in a similar way.

thomasjpfan committed:
Apply suggestions from code review

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into target_regression_encoder_v2_pr

thomasjpfan committed:
FIX Fixes compile issue

thomasjpfan committed:
CLN Removes pow

thomasjpfan committed:
CLN Use nogil and pointers

thomasjpfan commented on sklearn/preprocessing/_target_encoder_fast.pyx:
```diff
@@ -0,0 +1,136 @@
+                current_encoding[cat_idx] = y_mean
+            else:
+                current_encoding[cat_idx] = lambda_ * means[cat_idx] + (1 - lambda_) * y_mean
+        encodings.append(current_encoding)
```
The overall data structure is a "jagged array" because each categorical feature may have different cardinalities. I updated this PR and kept the Python list to store the ndarrays and a `double**` to have access to the pre-allocated encodings without the gil.

(I considered using C++ vector instead of `double**`, but this seems simple enough to stick with `malloc/free` and not require a C++ compiler)

ogrisel replied:
I see you changed your mind and decided to use a vector instead in 665e2e9. I suppose this makes the memory management cleaner as I assume that Cython will do the ref counting automatically for the returned datastructure.

thomasjpfan replied:
In both implementations, Python is handling the reference counting of the output. For C++, there is a temporary `vector[double *]` that points to the outputs. For C, there was a `double**` that points to the outputs. I do not think setting pointers to the output data would increment the Python reference count, because pointers are not PyObjects.

The biggest change with C++ is that `vector[double*]` is [RAII](https://en.cppreference.com/w/cpp/language/raii) and has it's own C++ reference count and freed when it gets out of scope.

thomasjpfan committed:
MNT Use vector instead

thomasjpfan committed:
DOC Adds note about missing values

thomasjpfan committed:
API Adds shuffle and random_state

thomasjpfan committed:
STY Slight reformatting

thomasjpfan committed:
Check random seeds [all random seeds]
test_fit_transform_not_associated_withy_if_ordinal_categorical_is_not

jovan-stojanovic commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -0,0 +1,337 @@
+    the global target mean with the target mean conditioned on the value of the
+    category. [MIC]_
+
+    :class:`TargetEncoder` considers missing values, such as `np.nan` or `None`,
+    as another category and encodes them like any other category.
```
> as of now, TargetEncoder has no explicit support for missing values and in particular encoded with `np.nan` in the original data are target encoded as any other categories.

In dirty-cat we have two parameters that handle missing values seen during fit and unknown categories seen during transform.

For missing values during fit, we added a parameter [`handle_missing`](https://github.com/dirty-cat/dirty_cat/blob/ea009b9/dirty_cat/_target_encoder.py#L49) that either imputes all the missing values as blank strings ' ' or raises an error. 

For unseen categories in transform, we have a [`handle_unknown`](https://github.com/dirty-cat/dirty_cat/blob/ea009b9/dirty_cat/_target_encoder.py#L43) parameter that either raises an error or encodes them with the prior mean of the target variable. This is mentionned in Micci-Barreca (2001) (page 30, first paragraph):
> "If missing values have no particular relationship with the dependent then S_{o} is going to converge towards a prior probability of the target which corresponds to a "neutral" representation of the missing value"
This was inspired by the `handle_missing` parameter of the [OneHotEncoder](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html) so I think users would easily understand this if you add it.

From what I see, you implement both under the hood, I guess that's what you mean @ogrisel by explicit support.
If needed to add the docstring, I think the one from dirty-cat could be adapted easily.

thomasjpfan replied:
A `handle_missing` parameter makes sense to me. Given the current size of this PR, I prefer to consider the `handle_missing` parameter in a follow up PR.

jovan-stojanovic commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -0,0 +1,337 @@
+            The type of target inferred with `"auto"` may not be the desired target
+            type used for modeling. For example, if the target consistent of integers
+            between 0 and 100, then :func:`~sklearn.utils.multiclass.type_of_target`
+            will infer the target as `"multiclass"`. In this case, setting
+            `target_type="continuous"` will the target as a regression problem. The
```
I guess there is a word missing?
```suggestion
            will infer the target as `"multiclass"`. In this case, setting
            `target_type="continuous"` will understand the target as a regression problem. The
```

jovan-stojanovic commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -0,0 +1,337 @@
+    n_features_in_ : int
+        Number of features seen during :term:`fit`.
+
+    feature_names_in_ : ndarray of shape (`n_features_in_`,)
+        Names of features seen during :term:`fit`. Defined only when `X`
+        has feature names that are all strings.
```
Is this redundant with the categories_ attribute? Seems to me that the output will be identical.

thomasjpfan replied:
They give different information. `feature_names_in_` are the column names of the input dataframe and `categories_` are the actual categories.  For example:

```python
from sklearn.preprocessing import TargetEncoder
import pandas as pd

X_df = pd.DataFrame({"pet": ["dog", "cat", "snake"]})
y = [0, 1, 0]
enc = TargetEncoder().fit(X_df, y)

enc.feature_names_in_
# array(['pet'], dtype=object)

enc.categories_
# [array(['cat', 'dog', 'snake'], dtype=object)]
```

jovan-stojanovic replied:
Ok, I see, thanks, I was mixing features and categories.

jovan-stojanovic reviewed (commented):
Hi, I am a dirty-cat maintainer and I wanted to give some insights based on our experience with the [TargetEncoder](https://dirty-cat.github.io/stable/generated/dirty_cat.TargetEncoder.html#dirty_cat.TargetEncoder).
Super nice to see this included in scikit-learn!

thomasjpfan committed:
API Change attriubte name to target_mean_

thomasjpfan committed:
Apply suggestions from code review

Co-authored-by: Jovan Stojanovic <62058944+jovan-stojanovic@users.noreply.github.com>

thomasjpfan committed:
STY Linting issue

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into target_regression_encoder_v2_pr

thomasjpfan commented:
Here are some recent updates:

- Changed `encoding_mean_` to `target_mean_`: [`a327b52` (#25334)](https://github.com/scikit-learn/scikit-learn/pull/25334/commits/a327b52f82101566c4bbfdec1ad6840d7eb1954a)
- Added `shuffle` and `random_state`: [`70f99fe` (#25334)](https://github.com/scikit-learn/scikit-learn/pull/25334/commits/70f99fe3c79084292ef70b9463bba5d4fa90af8b)
- Tested with `[all random seeds]` with the new `global_random_seed` test which passes in the [CI](https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=52635&view=logs&j=7d13af21-9cb9-5d40-483b-ea0f074409c6&t=97bcaf17-ac30-535a-f717-a6f6663c5884)

ogrisel commented on sklearn/preprocessing/_target_encoder_fast.pyx:
```diff
@@ -0,0 +1,165 @@
+cimport numpy as cnp
+import numpy as np
+
+cnp.import_array()
```
Do we really need this here? It seems that we only use typed memory views in this code. Or maybe this is necessary to call `np.empty`?

thomasjpfan replied:
`cnp` is still used for the dtypes and I think it's safer to include whenever we cimport `cnp`. I commented about this recently in https://github.com/scikit-learn/scikit-learn/pull/25780#issuecomment-1460865458 and da-woods from Cython recommends it: https://groups.google.com/g/cython-users/c/6qGmzvIv7Io

ogrisel replied:
Indeed:

> My summary recommendation is: if you cimport Numpy the safest thing is to always call import_array(). The cases where it isn't needed depend on Cython version (and possibly Numpy version) so that's why I'm recommending that it's always needed. I do plan to go through the Cython docs and update it.

Maybe @jjerphan could be expanded on this topic :)

jjerphan replied:
Based on remark of Cython developers, we agreed that calling `cnp.import_array` is safer.

---

FYI, this is something we recently have discussed in https://github.com/scikit-learn/scikit-learn/pull/25780.

Side-note: some calls to `cnp.import_array` might have been removed before this discussion.

ogrisel commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -0,0 +1,340 @@
+
+    >>> # On the other hand, a low `smooth` parameter puts more weight on target
+    >>> # conditioned on the value of the categorical:
+    >>> enc_no_smooth = TargetEncoder(smooth=1.0).fit(X, y)
+    >>> enc_no_smooth.encodings_
+    [array([20..., 80..., 43...])]
```
I think this should be updated to demonstrate the `smooth=0.0` edge case once the division by zero bug has been addressed.

Or alternatively the variable should be renamed to `enc_low_smooth` instead of `enc_no_smooth`.

ogrisel commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -0,0 +1,340 @@
+        "categories": [StrOptions({"auto"}), list],
+        "target_type": [StrOptions({"auto", "continuous", "binary"})],
+        "smooth": [StrOptions({"auto"}), Interval(Real, 0, None, closed="left")],
+        "cv": [Interval(Integral, 2, None, closed="left")],
```
Shall we ally to pass `cv=None` to disable internal CV? This could be useful for educational purposes and statistically study the impact of the nested CV.

ogrisel replied:
As discussed at the triage meeting, we can actually manually investigate the impact of disabling the cross-validation by manually calling `.fit(X_train).transform(X_train)` in our own code snippets. But this prevents using a pipeline to study the impact of the internal cross-validation. But maybe this is enough.

thomasjpfan replied:
Looking at this again, I am +0.3 with adding `cv=None` for educational purposes. To include `cv=None`, I would want more docs explaining why `cv=None` is normally bad to do. Given the current scope of this PR, can we push `cv=None` for a follow up PR?

I recall you wanted to add a test with `cv=None`. In that case, we can have a "hidden" `None` option that disables the CV, used only for testing purposes.

lorentzenchr replied:
Yes. Let‘s keep this PR scoped and postpone `cv=None` to a possible future PR.

ogrisel replied:
I am also +0.42 for adding `cv=None` and +1 for doing this in a follow-up PR along with an (updated or dedicated) example that document the pitfall of using `cv=None`.

ogrisel replied:
If we implement `cv=None` we can make the test in 26d2429 more natural.

ogrisel commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -0,0 +1,368 @@
+    assert_allclose(X_test_trans, np.repeat([[y_mean]], 2, axis=0))
+
+
+def test_fit_transform_not_associated_withy_if_ordinal_categorical_is_not(
```
```suggestion
def test_fit_transform_not_associated_with_y_if_ordinal_categorical_is_not(
```

ogrisel reviewed (commented):
Here is another pass of review. In particular I found a new bug with `smooth=0` with rare categories (e.g. a category that appears only once in the training set).

I am also working on an additional test (that should pass, so no discovered bug) that I plan to push directly to the PR and then we can discuss its value there.

I think it would be interesting to be able to disable the nested cv by passing `cv=None` to be able to study the impact of this for educational purposes as commented below.

This would probably make it possible to write even more interesting tests or examples.

thomasjpfan committed:
FIX Fixes edge case with smooth=0.0 and unknown categories in cv

thomasjpfan committed:
Apply suggestions from code review

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into target_regression_encoder_v2_pr

thomasjpfan commented on sklearn/preprocessing/_target_encoder_fast.pyx:
```diff
@@ -0,0 +1,165 @@
+                counts[X_int_tmp] += 1.0
+
+            for cat_idx in range(n_cats):
+                encoding_vec[feat_idx][cat_idx] = sums[cat_idx] / counts[cat_idx]
```
Fixed and added test in [`1ea0407` (#25334)](https://github.com/scikit-learn/scikit-learn/pull/25334/commits/1ea0407608063f51d0d523c468e64be08a35e515)

ogrisel reviewed (commented):
Some more pending review comments. I still want to improve the tests by adding non-regression statistical integration tests. Will do that on Monday hopefully.

thomasjpfan committed:
TST Improves testing maintainability

ogrisel committed:
TST add statistical-integration tests to check the benefit of internal CV

ogrisel commented:
I pushed 26d2429 to add a statistical non-regression test that ensures that the nested cross-validation in `fit_transform` is actually useful to in a minimal synthetic dataset that is hopefully not to artificial but reflects situations that can happen in practice.

I am still no decided whether this should better be a pitfall style example or a test. I think having it in a test makes it less likely to introduce a change in this encoder that would cause a regression. But maybe the existing tests are enough.

I am curious to hear your feedback.

ogrisel commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -0,0 +1,341 @@
+    smooth : "auto" or float, default="auto"
+        The amount of mixing of the categorical encoding with the global target mean. A
+        larger `smooth` value will put more weight on the global target mean.
+        If `"auto"`, then `smooth` is set to an empirical Bayes estimate.
```
I think we should allow for an `n_features` length parametrization here (e.g. list and/or dict with feature names as keys): the optimal smoothing for one feature is not necessarily optimal for others.

This can be done as follow-up PR though. It feels border-line YAGNI to me. In the mean time it possible to use a column-transformer to configure a per-feature `TargetEncoder` instance with dedicated `smooth` values per features.

ogrisel commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -0,0 +1,548 @@
+    model_no_cv = linear_regression.fit(X_enc_no_cv_train, y_train)
+
+    # The linear regression model should always overfit because it assigns
+    # all its weight to the extremely high cardinality feature and completely
+    # ignores the informative feature:
```
```suggestion
    # too much too much weight to the extremely high cardinality feature relatively
    # to the informative feature:
```

ogrisel committed:
Fix comment in test to more accurately describe what's happening

thomasjpfan committed:
TST Use RNG for permutation

ogrisel reviewed (approved):
Looks good to me. Based on `test_target_encoding_for_linear_regression` I think we need to do the nested cross-val by default and break the usual `fit_transform` <=> `fit + transform` implicit equivalence of other scikit-learn estimators. In particular, I don't see how to compute the "real" training accuracy: to do so we would need a `fit_score` method on pipelines (which could be a good idea by the way to save some redundant computation, but this is a digression).

Anyways, I don't see any other way around, and to me the protection against catastrophic overfitting caused by noisy high-cardinality categorical features outweighs the potentially surprising (but well documented) behavior of `fit_transform`.

thomasjpfan commented:
> I am still no decided whether this should better be a pitfall-style example or a test.

I think both are useful. I have not seen an example similar to your test case that demonstrates why the internal validation is useful.

In a follow up PR, we can convert the test into a pitfall style example and link it in the docstring for `cv`. (I'm still unsure about adding `cv=None`)

ogrisel commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -0,0 +1,548 @@
+    assert (np.abs(coef[1:]) < 0.2).all()
+
+    # Let's now disable the internal cross-validation by calling fit and then
+    # transform separately on the training set, even when using the empirical
+    # Bayes smoothing:
```
```suggestion
    # transform separately on the training set:
```

ogrisel commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -0,0 +1,548 @@
+
+    # The linear regression model should always overfit because it assigns
+    # too much weight to the extremely high cardinality feature relatively to
+    # the informative feature:
```
```suggestion
    # the informative feature. Note that this is the case even when using
    # the empirical Bayes smoothing which is not enough to prevent such
    # overfitting alone.
```

ogrisel reviewed (commented):
Spotted another typo in the inline comment of the new test.

ogrisel committed:
More fixes in inline comments for the new test

ogrisel commented:
@lorentzenchr @betatim @glemaitre any more feedback?

@jovan-stojanovic you might be interested in the new test: I checked that dirty_cat's `TargetEncoder` can also lead to this kind of overfitting problems because of the lack of built-in CV during `fit_transform`.

ogrisel commented:
Another pitfall I discovered when experimenting with this PR:

If you have a mix of informative and non-informative categorical features (e.g. `f_i` and `f_u`), and if you apply `StandardScaler()` after `TargetEncoder()` then you can run in big troubles: the variance of `target_encoder.fit_tramsform(X_train[[f_u]])` can be very small but non-zero, so `StandarScaler` can rescale this massively which is probably a bad idea for a downstream distance based model (e.g. a kernel machine, nearest neighbors and so on).

However if you use the raw target encoded values of `f_i` and `f_u` in conjunction of true numerical features that have all been scaled (via a dedicated entry in a column transformers) and if y has a much larger standard deviation, then `f_i` will have a very large standard deviation compared to the independently scaled numerical features, which is problematic.

I see two possible solutions:

- `a)` add an option to `TargetEncoder` to internally use a scaled version of `y_train` to compute the encodings;
- `b)` make it possible for StandardScaler or those scalers to use shared mean and scale statistics for a group of columns:

```python
preprocessor = ColumnTransformer(
    [
        (
            "categorical",
             make_pipeline(TargetEncoder(), StandardScaler(shared_scale=True),
            ["f_i", "f_u"],
        ),
    ],
    remainder=StandardScaler(),
)
```

Option `b)` is probably useful beyond post-processing target encoded categorical values. For instance @TomDLT and I encountered a similar problem a long time ago when running `make_pipeline(StandardScaler(), LogisticRegression())` with the saga solver on MNIST: some pixel values are nearly constant 0 or 1 and having per-feature scaling is catastrophically hurting the condition number of `X.T @ X` if I recall correctly. At the time we just decided to use `MinMaxScaler` instead and move along. But this is a real problems when groups of features should be re-scaled together to preserve meaning (e.g. pixel intensities, EEG channels, MEL features for audio signal and so on).

Even if we decide would also be `a)` convenient UX improvement, we can delay this to a follow-up PR.

I just wanted to brain-dump this here so that we can think about it when we work on a pitfall example for `TargetEncoder`-based feature engineering.

/cc @jovan-stojanovic who might also be interested for dirty_cat.

thomasjpfan commented:
> so overall, I am not 100% sure the weakref hack would be a usability improvement or not.

I think it borders on being too magical. For example, if the data is sliced the same way or copied, the references are not the same:

```python
import numpy as np
import weakref

X = np.random.randn(10, 10)

X1 = X[:4]
X2 = X[:4]
X3 = X1.copy()

X1_ref = weakref.ref(X1)

assert X1 is X1_ref()
assert X2 is not X1_ref()
assert X3 is not X1_ref()
```

For reference, cuML's TargetEncoder holds the [training data and checks all the values](https://github.com/rapidsai/cuml/blob/d41a081cc5809ef0f14e1c7f4f77749ab044ec93/python/cuml/preprocessing/TargetEncoder.py#L431-L446).

> a) add an option to TargetEncoder to internally use a scaled version of y_train to compute the encodings;

At one point, I had something similar implemented in https://github.com/scikit-learn/scikit-learn/pull/17323 as the default. I think it's reasonable to use a scaled version of the target for encoding purposes.

TomDLT commented:
> For instance @TomDLT and I encountered a similar problem a long time ago when running `make_pipeline(StandardScaler(), LogisticRegression())` with the saga solver on MNIST: some pixel values are nearly constant 0 or 1 and having per-feature scaling is catastrophic.

Yes, some MNIST pixels reach a value of 250 after rescaling (see Details below), which is quite  out of distribution for a unit-norm Normal distribution. We should add a warning when some outputs of `StandardScaler` are large (say above 100, or even 10).

> This is a real problems when groups of features should be re-scaled together to preserve meaning (e.g. pixel intensities, EEG channels, MEL features for audio signal and so on).

Yes, this is a big limitation of `StandardScaler`. +1 for adding `StandardScaler(shared_mean=True, shared_scale=True)`.

<details>

![Figure_1](https://user-images.githubusercontent.com/11065596/225419937-1192497f-8844-434d-afa1-b38385b1ebc4.png)

```py
import matplotlib.pyplot as plt
from matplotlib.colors import LogNorm

from sklearn.datasets import fetch_openml
from sklearn.preprocessing import StandardScaler

mnist = fetch_openml("mnist_784", as_frame=False, parser="pandas")
X, y = mnist.data, mnist.target

X_scaled = StandardScaler().fit_transform(X)
max_values = X_scaled.max(axis=0)

fig, ax = plt.subplots()
image = ax.imshow(max_values.reshape(28, 28), cmap=plt.get_cmap("viridis", 6),
                  norm=LogNorm())
ax.set(xticks=[], yticks=[], title="Maximum value of each scaled MNIST feature")
fig.colorbar(image)
plt.show()
```

ogrisel commented:
> For reference, cuML's TargetEncoder holds the [training data and checks all the values](https://github.com/rapidsai/cuml/blob/d41a081cc5809ef0f14e1c7f4f77749ab044ec93/python/cuml/preprocessing/TargetEncoder.py#L431-L446).

Note that we could use a weakref + a concrete value check. But even that would feel to complex/magical. +0.5 for keeping the code as it is.

glemaitre commented on examples/preprocessing/plot_target_encoder.py:
```diff
@@ -0,0 +1,227 @@
+# be a reviewer:
+from sklearn.datasets import fetch_openml
+
+wine_reviews = fetch_openml(data_id=42074, as_frame=True, parser="liac-arff")
```
This should be fixed now.

```suggestion
wine_reviews = fetch_openml(data_id=42074, as_frame=True, parser="pandas")
```

glemaitre reviewed (approved):
I have only one nitpick.

ogrisel committed:
Use pandas to parse the fetched the example dataset.

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

lorentzenchr commented:
Just an idea: For detection of the training set during `transform`, we could store mean, variance and maybe 2 quantiles. This should be pretty unique.

ogrisel commented:
The checks need to happen before encoding on the categorical variables. We could store the per feature category counts instead. Maybe with a few random probe records.

ogrisel commented:
The checks need to happen before encoding on the categorical variables. We could store the per feature category counts instead. Maybe with a few random probe records.

But this would be quite catastrophic in case of false positives.

BenjaminBossan commented:
Pretty nice addition, thanks for this.

A small question: According to the tags, this does not require `y`, should this be changed?

```python
>>> from sklearn.preprocessing import TargetEncoder
>>> from sklearn.utils._tags import _safe_tags
>>> _safe_tags(TargetEncoder())['requires_y']
False
```
